node/src/actors/chain_manager/handlers.rs

```diff
@@ -1292,20 +1293,43 @@ impl Handler<GetDataRequestInfo> for ChainManager {
 }
 
 impl Handler<GetBalance> for ChainManager {
-    type Result = Result<u64, failure::Error>;
+    type Result = ResponseActFuture<Self, Result<(u64, u64), failure::Error>>;
 
     fn handle(&mut self, GetBalance { pkh }: GetBalance, _ctx: &mut Self::Context) -> Self::Result {
         if self.sm_state != StateMachine::Synced {
-            return Err(ChainManagerError::NotSynced {
-                current_state: self.sm_state,
-            }
-            .into());
+            return Box::pin(actix::fut::err(
+                ChainManagerError::NotSynced {
+                    current_state: self.sm_state,
+                }
+                .into(),
+            ));
         }
 
-        Ok(transaction_factory::get_total_balance(
-            &self.chain_state.unspent_outputs_pool,
-            pkh,
-        ))
+        let magic = self.get_magic();
+
+        let res = storage_mngr::get::<_, ChainState>(&storage_keys::chain_state_key(magic))
+            .into_actor(self)
+            .then(move |chain_state_from_storage, act, _| {
+                let result = match chain_state_from_storage {
+                    Ok(Some(x)) => transaction_factory::get_total_balance(
```

Comment:
The `Ok(Some(x))` and `Ok(None)` branches are the same, so maybe they can be merged as:

```suggestion
                    Ok(x) => transaction_factory::get_total_balance(
```